README
What is gld ?

	gld stands for GreyList Daemon.
	gld is a standalone policy delegation server for postfix that implements
	the greylist algorithm as defined at http://www.greylisting.org

What are the requirements ?

	gld needs the following software:
		- a postfix 2.1 or higher server
		- a mysql server or a pgsql server

How to build gld ?

	This should be easy, just run ./configure and then make and wait ...
	Of course you need to have include files and libraries for MySQL support.

	If the compilation fails with mysql related errors, try to run 
	./configure --with-mysql=DIR .


How to install gld ?

	Just run make install

	The install copy only 2 files into your system.
	The files are installed as follow:

	/etc/gld.conf.sample   : the sample config file
	/usr/local/bin/gld     : the gld executable file


	Afterward, edit /etc/gld.conf.sample file and modify it to suit your needs.
	Then rename it to gld.conf .

	Please, PLEASE read the README-SECURITY file before choosing your options.

	Then, create the tables whitelist and greylist on your mysql server
	I have provided the script tables.sql that helps you to do this task.

	Then start gld and see if it run or if an error message is displayed

	If you are here, you are done, now just configure postfix to use
	your brand new greylist server .

	To configure postfix just add the following line:

	    check_policy_service inet:127.0.0.1:2525

	to the end of your smtpd_recipient_restrictions statement of your main.cf file.
	You don't have to edit or modify your master.cf file .


How to start gld ?

	just run gld without any parameters.


How to stop gld ?

	Just kill the process with the TERM signal
	(SIGTERM is the default signal sent by the kill command)


How to reload the configuration of the server ?

	Just send the HUP signal to the server
	ie: kill -HUP <pid> or killall -HUP gld


How to know the version of gld ?

	Just try: gld -v
	

What happens if the mysql server goes down ?

	Depending on the configuration of gld, the server can refuse to answer
	or send a 'dunno' response, which will likely accept the mails.
	Please read gld.conf for more informations.


What happens if gld goes down ?

	Well, gld *MUST* be running while postfix is running.
	make sure you start gld before you start postfix .

	If gld should die (this has never occured here)
	postfix will not be able to connect to the policy server
	and will return a '450 server configuration' .

	Thus, you won't be able to receive any greylisted mail .....


What king of replies this server send to postfix ?

	gld sends only 2 replies to postfix.

	if the email is to be refused, then gld reply: action=defer_if_permit MESSAGE
	otherwise gld reply: action=dunno


I want to greylist only some emails and only some domains, How to do ?

	postfix has a cool feature for that named policy maps.
	Let say you want to greylist only the email grey@foo.bar and the whole domain bar.com
	here follow hos to set up postfix for that.

	First, define a policy map named greylist_policy in main.cf
	to do this just add the following lines in main.cf :

	smtpd_restriction_classes = greylist_policy
	greylist_policy = check_policy_service inet:127.0.0.1:2525

	Then in the end of  smtpd_recipient_restrictions
	add the following line: check_recipient_access hash:/etc/postfix/A_FILE_OF_YOUR_CHOICE
	instead of the standard line: check_policy_service inet:127.0.0.1:2525

	Finally create a text file named /etc/postfix/A_FILE_OF_YOUR_CHOICE
	which contains domain and emails to be greylisted
	the format is:
		email greylist_policy
		domain greylist_policy

	thus in our example, the file would contain
		grey@foo.bar greylist_policy
		bar.com greylist_policy

	Finally make this text file a hash database with the command: postmap /etc/postfix/A_FILE_OF_YOUR_CHOICE
	and you are done ! .



Where do I report bugs,suggestions,insults ?

	Just send a mail to salim@gasmi.net
	(of course this email is greylisted ..)
	But please, include the version of gld you use and the OS you are running.


Note from the author:

	I want to thanks all people who mailed me with suggestions, patches
	and especially the following guys:
		- Lefteris Tsintjelis for his help and support for testing 1.6.
		- Dietmar Braun for his idea of the training mode.
		- Wayne Smith for his support and suggestions.
		- Santiago Vila for maintaining the Debian package.
		- Blaz Zupan for maintaining the FreeBSD port.
		- Volker Tanger for the nice howto.
		- Brian Truelsen for his help and suggestions on MXGREY.
		- Wietse Venema for postfix.



HISTORY
01/05/2006 : V 1.7

	- Removed the KEEPDBOPEN option as it could lead to multiple instance of gld running
	  when mail traffic is low and thus father process MYSQL connection timeout .
	  Furthermore, after tests, efficency is the same .
	 
	  KEEPDBOPEN option is now silently ignored to forced to NO .


16/09/2005 : V 1.6

	- Removed the algorithm lightgreydomain as the new mxgrey does better and simpler

	- Removed the UPDATE option, now gld always update triplets .

	- Fixed a minor flaw in the MXGREY algorithm.

	- Now you can configure MXGREY to consider an ip as a safe MX after X succesful
	  greylists instead of only 1 . (read gld.conf for details)

	- Now gld updates the counters only when mail is not greylisted

	- Added Training mode, read gld.conf for details

	- Now gld verify that if you supply a custom smtp code, it's a 4XX code
	  otherwise gld discard it and use 450 .

	- If gld cannot connect to MySQL server on startup
	  it will not refuse to run anymore, but will set keepdbopen to 0 and accept to run .

	- The sample config file provided now only listen to loopback
	  only accept connection from localhost and runs as nobody/ nobody.

	WARNING: if you were using lightgreydomain algorithm, it has been discontinued
	use MXGREY in place, please read gld.conf for details.


21/04/2005 : V 1.5.2

	- Fixed a bug in MXGREY algorithm


13/04/2005 : V 1.5

	- Fixed several possible buffer overflow by removing all sprintf() calls
	and replacing them by snprintf()

	- Added pgsql support, use configure --with-pgsql to compile
	gld with pgsql support

	- Added the MXGREY parameter, please read gld.conf for details.

	- Added KEEPDBOPEN parameter, please read gld.conf for details.

	- Added Configurable SMTP return code 

	- Updated the white list to the lastest known.


19/08/2004 : V 1.4

	- Added the USER and GROUP parameters to allow running gld
	  as a specific user/group instead of root.

	- Added the LIGHTGREYDOMAIN parameter for greylisting on domains
	  please read gld.conf for details.

	- Added CLIENTS parameter, to restrict clients allowed to connect to gld.


15/08/2004 : V 1.3.1

	- Parameters are now checked to avoid a possible buffer overflow

	- Parameters are now checked to avoid a possible SQL injection

	- Fixed the bug which makes gld crash when displaying infos for an empty database


19/07/2004 : V 1.3

	- Now, when LIGHTGREY is active, full ip+network is checked in whitlists
	  instead of the network only.

	- Added the LOOPBACKONLY parameter, to make gld listen only to loopback (127.0.0.1) .

	- Added the DNSW parameter for DNS whitelist support (cf DNSWL in gld.conf)

	- Added the FACILITY parameter to define the syslog facility we want to use

	- Added the -C switch to simulate what -c would do

	- Added the -k and -K switchs to clean the one hits entries in the database

	- Added the -i switch to display some database informations

	- Added the -d switch to enable debug mode

	- Added in the documentation a mini howto to setup postfix to greylist only some emails/domains.

	
25/06/2004 : V 1.2

	- Improved child management to avoid zombies on some BSD.
	  Thanks to Michael Rasmussen <mir@datanom.net> that solved the bug that caused 
	  gld to leave spawned children as zombies on BSD.
	
	- Added the light greylisting algorithm (cf LIGHTGREY parameter in gld.conf).

	- Added the -c option to gld to clean up the database (try gld -h).

	- All options have now a default value and are no more mandatory.

	- gld now use a configure script and should be easier to compile on some unixes.

	- Improved documentation and given a postfix configuration example.


20/06/2004 : V 1.1

	- Some tries, this version has never been released.


10/05/2004 : V 1.0

	- Initial release.


gld.h
#include <stdio.h>
#include <stdlib.h>
#include <fcntl.h>
#include <unistd.h>
#include <string.h>
#include <time.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <sys/socket.h>
#include <sys/ipc.h>
#include <sys/shm.h>
#include <sys/types.h>
#include <sys/wait.h>

#include <pwd.h>
#include <grp.h>

#ifdef HAVE_NETDB_H
#include <netdb.h>
#endif


#ifdef HAVE_SYS_TIME_H
#ifdef TIME_WITH_SYS_TIME
#include <sys/time.h>
#endif
#endif

#define TOUT    10			// Network timeout in seconds
#define BLEN    2048			// Our buffer size
#define QLEN    20480			// Our Query Size must be > than 6*BLEN
#define NLEN	32			// Max # of networks in our table.
#define REQ     "smtpd_access_policy"	// The string to be matched
#define CONF	"/etc/gld.conf"		// The default config file

#define MSGGREYLIST	0
#define MSGLOCALWL	1
#define MSGDNSWL	2

#define VERSION "1.7"

typedef struct network
{
	unsigned long netw;
	unsigned long mask;
}network;

typedef struct config
{
	int port;			// Port to listen to
	int maxcon;			// max # of connections
	long mini;			// Minimum time for greylist
	int syslog;			// Shall we write to the syslog
	int accept;			// Shall we return OK in case of error
	int whitelist;			// Shall we lookup the whitelist table
	int light;			// Shall we use light greylisting ?
	int facility;			// Syslog facility to use
	int loopback;			// Shall we bind only to loopback
	int mxgrey;			// Shall we use the mexgrey algorithm ?
	int debug;			// Shall we display debug informations
	int training;			// Shall we activate training mode
	char dnswl[512];		// The domain to use if we do DNSWL
	char message[512];		// The text we display
	char sqlhost[128];		// SQL server
	char sqluser[128];		// SQL User
	char sqlpasswd[128];		// SQL password
	char sqldb[128];		// SQL Database name
	char user[128];			// The user we setuid to
	char grp[128];			// The group we setgid to
	network nets[NLEN];		// The networks allowed to connect
	int nbnet;			// # of networks in nets
}config;

//
// Prototypes
//

int ReadConfig(char *file,config *conf);
int HandleChild(int s,config *cnf);
void TheEnd(int s);
int SQLConnect(char *host,char *user,char *passwd,char *db);
void SQLClose(void);
long SQLQuery(char *q);
int GreyList(char *ip,char *sender,char *recipient,config *conf);
void Log(config *conf,char * recipient,char *sender,char *ip,int white);
void ErrorLog(config *conf,char *msg);
void Reload(int s);
void ShowBaseInfo(void);
void Quote(char *str);
int ReadClients(config *conf,char *str);
unsigned long Hash(char *ip);
int CheckIP(config *conf,char *ip);
unsigned long CidrMsk(int msk);
int MyDaemon(int nochdir, int noclose);


sql.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#ifdef _DB_MYSQL_
#include <mysql/mysql.h>
#endif
#ifdef _DB_PGSQL_
#include "libpq-fe.h"
#endif
#include "gld.h"

#ifdef _DB_MYSQL_
static int MySQLConnect(char *host,char *user,char *passwd,char *db);
static void MySQLClose(void);
static long MySQLQuery(char *q);
#endif
#ifdef _DB_PGSQL_
static int PgSQLConnect(char *host,char *user,char *passwd,char *db);
static void PgSQLClose(void);
static long PgSQLQuery(char *q);
#endif

int SQLConnect(char *host,char *user,char *passwd,char *db)
{
#ifdef _DB_MYSQL_
return(MySQLConnect(host, user, passwd, db));
#endif
#ifdef _DB_PGSQL_
return(PgSQLConnect(host, user, passwd, db));
#endif
}

void SQLClose(void)
{
#ifdef _DB_MYSQL_
return(MySQLClose());
#endif
#ifdef _DB_PGSQL_
return(PgSQLClose());
#endif
}

long SQLQuery(char *q)
{
#ifdef _DB_MYSQL_
return(MySQLQuery(q));
#endif
#ifdef _DB_PGSQL_
return(PgSQLQuery(q));
#endif
}

void ShowBaseInfo(void)
{
char query[QLEN];
int c;
long now;

now=time(0);

snprintf(query,sizeof(query)-1,"select count(*) from greylist");
c=SQLQuery(query);
printf("# of entries in the database         : %d\n",c);

if(c!=0)
	{
	snprintf(query,sizeof(query)-1,"select count(*) from greylist where n=1");
	c=SQLQuery(query);
	printf("# of one hit entries in the database : %d\n",c);

	snprintf(query,sizeof(query)-1,"select min(first) from greylist");
	c=SQLQuery(query);
	printf("Oldest entry in database             : %ld days ago\n",(now-c)/86400);
	}
}

void Quote(char *str)
{
int i,l;

l=strlen(str);
for(i=0;i<l;i++)
	if(str[i]=='"' || str[i]=='\'' || str[i]==';') str[i]=' ';

}

#ifdef _DB_MYSQL_

MYSQL *sql;

int MySQLConnect(char *host,char *user,char *passwd,char *db)
{

MYSQL *sqlh;

sqlh=mysql_init(NULL);
if(sqlh==(MYSQL *)NULL) return(-1);

sql=mysql_real_connect(sqlh,host,user,passwd,db,0,NULL,0);
if(sql==(MYSQL *)NULL) return(-2);

return(0);
}

void MySQLClose(void)
{
if(sql!=(MYSQL *)NULL) mysql_close(sql);
}

long MySQLQuery(char *q)
{
int r;
MYSQL_RES *result;
MYSQL_ROW row;
int n;
long ret;

if(sql!=(MYSQL *)NULL) r=mysql_query(sql,q); else r=-1;
if(r!=0) return(-1);

result=mysql_store_result(sql);
if(result==(MYSQL_RES *)NULL) return(0);

n=mysql_num_rows(result);
if(n==0) return(0);

row=mysql_fetch_row(result);
ret=atol(row[0]);
mysql_free_result(result) ;

return(ret);
}

#endif
#ifdef _DB_PGSQL_

PGconn *sql;

int PgSQLConnect(char *host,char *user,char *passwd,char *db) {
	char conninfo[1024];

	snprintf(conninfo,sizeof(conninfo)-1,"host = %s user = %s password = %s dbname = %s", host, user, passwd, db);
	sql = PQconnectdb(conninfo);

	if (PQstatus(sql) == CONNECTION_BAD) {
		fprintf(stderr, "Connection to database \"%s\" failed.\n", db);
		fprintf(stderr, "%s", PQerrorMessage(sql));
		PQfinish(sql);
		return(-2);
	}

	return(0);
}

void PgSQLClose(void) {
	PQfinish(sql);
	return;
}

long PgSQLQuery(char *q) {
	PGresult *result;
	long ip;

	if (sql != NULL) {
		result = PQexec(sql, q);
	} else {
		return(-1);
	}

	if (PQntuples(result) == 0) {
		return(0);
	}

	ip = atol(PQgetvalue(result, 0, 0));
	PQclear(result);
	return(ip);
}

#endif


server.c
#include "gld.h"
#include "sockets.h"

TcpServer srv;
config conf;

int main(int argc,char **argv)
{
int s;
int cid;
int c;
char clean=0;
char query[QLEN];
int status;
struct group *grp;
struct passwd *user;

if(argc==2 && strcmp(argv[1],"-v")==0)
	{
	printf("gld %s <salim@gasmi.net> <http://www.gasmi.net>\n",VERSION);
	exit(0);
	}

if(argc==3 && strcmp(argv[1],"-c")==0) clean=1;
if(argc==3 && strcmp(argv[1],"-C")==0) clean=2;
if(argc==3 && strcmp(argv[1],"-k")==0) clean=3;
if(argc==3 && strcmp(argv[1],"-K")==0) clean=4;
if(argc==2 && strcmp(argv[1],"-i")==0) clean=5;

if(argc!=1 && clean==0 && strcmp(argv[1],"-d")!=0)
	{
	printf("Usage: gld [-c <n>|-C <n>|-k <n>|-K <n>|-h|-v]\n");
	printf(" gld -c <n> : clean the database for ALL entries not updated since <n> days\n");
	printf(" gld -C <n> : show what the -c option would do, without doing it\n");
	printf(" gld -k <n> : clean the database for entries not updated since <n> days with only one hit \n");
	printf(" gld -K <n> : show what the -k option would do, without doing it\n");
	printf(" gld -i     : show some database informations\n");
	printf(" gld -d     : enable debug mode\n");
	printf(" gld -v     : display version\n");
	printf(" gld -h     : display Usage\n");
	exit(1);
	}

if(ReadConfig(CONF,&conf)!=0)
	{
	printf("Invalid config file %s\n",CONF);
	exit(2);
	}

if(argc==2 && strcmp(argv[1],"-d")==0) conf.debug=1;

signal(SIGTERM,TheEnd);
signal(SIGHUP,Reload);
signal(SIGCHLD,NoZombies);

//
// Here we drop privileges and setuid/setgid if needed
//

if(conf.grp[0]!=0)
	{
	grp=getgrnam(conf.grp);
	if(grp==(struct group *)NULL)
		{
		printf("Group %s not found, please check the GROUP variable in gld.conf\n",conf.grp);
		exit(10);
		}

	if(setgid(grp->gr_gid)!=0)
		{
		printf("Unable to setgid to %s\n",conf.grp);
		exit(11);
		}

	if(conf.debug==1) printf("setgid to %s OK\n",conf.grp);
	}

if(conf.user[0]!=0)
	{
	user=getpwnam(conf.user);
	if(user==(struct passwd *)NULL)
		{
		printf("User %s not found, please check the USER variable in gld.conf\n",conf.user);
		exit(10);
		}

	if(setuid(user->pw_uid)!=0)
		{
		printf("Unable to setuid to %s\n",conf.user);
		exit(12);
		}

	if(conf.debug==1) printf("setuid to %s OK\n",conf.user);
	}


//
// Now we do what we have to do
//


if(clean!=0)
	{
	if(SQLConnect(conf.sqlhost,conf.sqluser,conf.sqlpasswd,conf.sqldb)<0)
		{
		printf("Unable to connect to MYSQL\n");
		exit(1);
		}

	if(clean==5)
		{
		ShowBaseInfo();
		SQLClose();
		exit(0);
		}

	if(clean==1 || clean==2)
	snprintf(query,sizeof(query)-1,"select count(last) from greylist where last < UNIX_TIMESTAMP()-86400*%d",atoi(argv[2]));

	if(clean==3 || clean==4)
	snprintf(query,sizeof(query)-1,"select count(last) from greylist where last < UNIX_TIMESTAMP()-86400*%d AND n=1",atoi(argv[2]));

        c=SQLQuery(query);

	if(clean==2 || clean==4)
		{
		printf("I would clean %d entries older than %d days\n",c,atoi(argv[2]));
		SQLClose();
		exit(0);
		}

	if(clean==1) snprintf(query,sizeof(query)-1,"delete from greylist where last < UNIX_TIMESTAMP()-86400*%d",atoi(argv[2]));
	if(clean==3) snprintf(query,sizeof(query)-1,"delete from greylist where last < UNIX_TIMESTAMP()-86400*%d and n=1",atoi(argv[2]));
        SQLQuery(query);
	SQLClose();

	printf("Cleaned %d entries older than %d days\n",c,atoi(argv[2]));
        exit(0);
	}


//
// Ok, here we start the server
//


srv=OpenTcpServer(conf.port,conf.maxcon,conf.loopback);
if(srv.sd==-1)
	{
	printf("Unable to bind to port %d\n",conf.port);
	perror("Error was: ");
	exit(1);
	}

if(conf.debug==1) printf("bind to port %d succesful\n",conf.port);

if(conf.syslog==1) ErrorLog(&conf,"gld started, up and running");
if(conf.debug==0) MyDaemon(0,0);
if(conf.debug==1) printf("Waiting for incoming connexions\n");

//
// The main loop
//
while(1==1)
	{
	s=WaitTcpServer(srv);
	if(s>=0)
		{
		cid=fork();
		if(cid < 0 && conf.syslog==1) ErrorLog(&conf,"Fork returned error code, no child");
		if(cid==0)
        		{
			c=HandleChild(s,&conf);
			if(c!=0 && conf.accept==1)
				WriteSocket(s,"action=dunno\n\n",14,TOUT);

			close(s);
			waitpid(-1, &status, WNOHANG);
			exit(0);
        		}
		close(s);
		}
	}

CloseTcpServer(srv);
exit(0);
}

int HandleChild(int s,config *cnf)
{
char buff[BLEN];
char request[BLEN];
char sender[BLEN];
char recipient[BLEN];
char ip[BLEN];
int n;
long ts;
int pid;

pid=getpid();

if(SQLConnect(cnf->sqlhost,cnf->sqluser,cnf->sqlpasswd,cnf->sqldb)<0)
	{
	if(cnf->debug==1) printf("%d: Unable to connect to MYSQL\n",pid);
	if(cnf->syslog==1)
		{
		snprintf(buff,sizeof(buff)-1,"Unable to connect to MYSQL\n");
		ErrorLog(cnf,buff);
		}
	return(-1);
	}

GetPeerIp(s,ip,buff);

//
// We check if this IP is authorized to connect to us
//

if(CheckIP(cnf,ip)!=1)
	{
	if(cnf->debug==1) printf("%d: Rejected New incoming connexion from %s (%s)\n",pid,buff,ip);
	if(cnf->syslog==1)
		{
		snprintf(buff,sizeof(buff)-1,"Rejected New incoming connexion from %s (%s)\n",buff,ip);
		ErrorLog(cnf,buff);
		}

	SQLClose();
	return(0);
	}

//
// Ok, The IP is accepted
//

if(cnf->debug==1) printf("%d: New incoming connexion from %s (%s)\n",pid,buff,ip);

ts=time(0);

request[0]=sender[0]=recipient[0]=ip[0]=0;

CloseTcpServer(srv);

while(1==1)
	{
	//
	// This functions does not read more than BLEN-1 bytes
	// from the network and thus no buffer overflow is possible
	//
	if(ReadLSocket(s,buff,BLEN-1,TOUT)<0)
		{
		if(cnf->syslog==1) ErrorLog(cnf,"Read Network error");
		if(cnf->debug==1) printf("%d: Read Network error\n",pid);
		SQLClose();
		return(-1);
		}

	//
	// To be sure that our buffer is null terminated to avoid
	// a buffer overflow, we manually set a null to the end of the buffer.
	//
	buff[BLEN-1]=0;

	//
	// Now, we are sure our buffer string length is no more than BLEN
	// as all parameters are defined also as buffers with a BLEN size
	// no buffer overflow is possible using strcpy .
	//

	if(strcmp(buff,"")==0) break;

	if(strncmp(buff,"request=",8)==0)
		strcpy(request,buff+8);

	if(strncmp(buff,"sender=",7)==0)
		strcpy(sender,buff+7);

	if(strncmp(buff,"recipient=",10)==0)
		strcpy(recipient,buff+10);

	if(strncmp(buff,"client_address=",15)==0)
		strcpy(ip,buff+15);

	}

//
// To be sure that our parameters are null terminated to avoid
// a buffer overflow, we manually set a null to the end of the parameters.
//

ip[BLEN-1]=0;
recipient[BLEN-1]=0;
sender[BLEN-1]=0;

//
// Then we remove nasty chars to avoid a possible SQL injection
//

Quote(ip);
Quote(recipient);
Quote(sender);

//
// Now, we can safely use, str** functions
//

if(sender[0]==0) strcpy(sender,"void@void");

if(strcmp(request,REQ)!=0 || recipient[0]==0 || ip[0]==0)
	{
	snprintf(buff,sizeof(buff)-1,"Received invalid data req=(%s) sender=(%s) recipient=(%s) ip=(%s)",request,sender,recipient,ip);
	if(cnf->syslog==1) ErrorLog(cnf,buff);
	if(cnf->debug==1) printf("%d: %s\n",pid,buff);
	SQLClose();
	return(-2);
	}

if(cnf->debug==1)
	printf("%d: Got the following valid data req=(%s) sender=(%s) recipient=(%s) ip=(%s)\n",pid,request,sender,recipient,ip);

n=GreyList(ip,sender,recipient,cnf);
if(cnf->debug==1) printf("%d: End of the greylist algo\n",pid);


if(n<0)
	{
	if(cnf->syslog==1) ErrorLog(cnf,"MySQL error");
	if(cnf->debug==1) printf("%d: MySQL error\n",pid);
	SQLClose();
	return(-3);
	}

if(n==0)
	{
        if(cnf->syslog==1) Log(cnf,recipient,sender,ip,MSGGREYLIST);
	if(cnf->debug==1) printf("%d: Decision is to greylist\n",pid);

	if(cnf->training==0)
		{
		if(atoi(cnf->message)<400 || atoi(cnf->message)>499)
		snprintf(buff,sizeof(buff)-1,"action=defer_if_permit %s\n\n",cnf->message);
		else snprintf(buff,sizeof(buff)-1,"action=%s\n\n",cnf->message);
		}
	else
		{
		if(cnf->debug==1) printf("%d: Training mode, sending dunno\n",pid);
		strcpy(buff,"action=dunno\n\n");
		}

        WriteSocket(s,buff,strlen(buff),TOUT);
        }
else
	{
	WriteSocket(s,"action=dunno\n\n",14,TOUT);
	if(cnf->debug==1) printf("%d: Decision is to not greylist\n",pid);
	}

SQLClose();
return(0);
}

void TheEnd(int s)
{
	int status;

	while(wait(&status) > 0);
	shutdown(srv.sd,2);
	CloseTcpServer(srv);
	exit(0);
}

void Reload(int s)
{
ReadConfig(CONF,&conf);
}

int MyDaemon(int nochdir, int noclose)
{
int fd;

switch (fork()) {
        case -1:
                return(-1);
        case 0:
                break;
        default:
                _exit(0);
}

if(setsid() == -1) return(-1);
if(!nochdir) (void)chdir("/");

if(!noclose && (fd = open("/dev/null", O_RDWR, 0)) != -1)
        {
        (void)dup2(fd, STDIN_FILENO);
        (void)dup2(fd, STDOUT_FILENO);
        (void)dup2(fd, STDERR_FILENO);
        if(fd>2) (void)close(fd);
        }
return(0);
}


cnf.c
#include "gld.h"

#ifdef HAVE_SYSLOG_H
#include <syslog.h>
#define FACSIZE 10

typedef struct facs {
    char *name;
    int fac;
}facs;

facs facys[] =
  {
    { "daemon", LOG_DAEMON },
    { "mail", LOG_MAIL },
    { "local0", LOG_LOCAL0 },
    { "local1", LOG_LOCAL1 },
    { "local2", LOG_LOCAL2 },
    { "local3", LOG_LOCAL3 },
    { "local4", LOG_LOCAL4 },
    { "local5", LOG_LOCAL5 },
    { "local6", LOG_LOCAL6 },
    { "local7", LOG_LOCAL7 }
  };
#endif

int ReadConfig(char *file,config *conf)
{
char *p;
char buffer[1024];
FILE *fic;
int i;

fic=fopen(file,"r");
if(fic==(FILE *)NULL) return(-1);

// We set the default values

strcpy(conf->sqlhost,"localhost");
strcpy(conf->sqluser,"myuser");
strcpy(conf->sqldb,"mydb");
strcpy(conf->sqlpasswd,"mypasswd");
strcpy(conf->message,"Greylisted");
conf->training=0;
conf->port=2525;
conf->maxcon=100;
conf->mini=60;
conf->syslog=1;
conf->accept=1;
conf->whitelist=1;
conf->light=0;
conf->dnswl[0]=0;
conf->loopback=1;
conf->debug=0;
conf->user[0]=0;
conf->grp[0]=0;
conf->nbnet=0;
conf->mxgrey=0;

#ifdef HAVE_SYSLOG_H
conf->facility=LOG_MAIL;
#endif

// Now we read

while(fgets(buffer,1024,fic)!=NULL)
        {
        p=(char *)strstr(buffer,"=");
        if(p!=NULL)
                {
		buffer[strlen(buffer)-1]=0;
                *p=0;
                if(strcmp(buffer,"CLIENTS")==0) ReadClients(conf,p+1);
                if(strcmp(buffer,"USER")==0) strcpy(conf->user,p+1);
                if(strcmp(buffer,"GROUP")==0) strcpy(conf->grp,p+1);
                if(strcmp(buffer,"DNSWL")==0) strcpy(conf->dnswl,p+1);
                if(strcmp(buffer,"SQLHOST")==0) strcpy(conf->sqlhost,p+1);
                if(strcmp(buffer,"SQLUSER")==0) strcpy(conf->sqluser,p+1);
                if(strcmp(buffer,"SQLDB")==0) strcpy(conf->sqldb,p+1);
                if(strcmp(buffer,"SQLPASSWD")==0) strcpy(conf->sqlpasswd,p+1);
                if(strcmp(buffer,"MESSAGE")==0) strcpy(conf->message,p+1);
                if(strcmp(buffer,"PORT")==0) conf->port=atoi(p+1);
                if(strcmp(buffer,"MAXCON")==0) conf->maxcon=atoi(p+1);
                if(strcmp(buffer,"TRAINING")==0) conf->training=atoi(p+1);
                if(strcmp(buffer,"MINTIME")==0) conf->mini=atol(p+1);
                if(strcmp(buffer,"MXGREY")==0) conf->mxgrey=atoi(p+1);
                if(strcmp(buffer,"SYSLOG")==0) conf->syslog=atoi(p+1);
                if(strcmp(buffer,"ERRACCEPT")==0) conf->accept=atoi(p+1);
                if(strcmp(buffer,"WHITELIST")==0) conf->whitelist=atoi(p+1);
                if(strcmp(buffer,"LIGHTGREY")==0) conf->light=atoi(p+1);
                if(strcmp(buffer,"LOOPBACKONLY")==0) conf->loopback=atoi(p+1);
		#ifdef HAVE_SYSLOG_H
                if(strcmp(buffer,"FACILITY")==0)
			for(i=0;i<FACSIZE;i++)
				if(strcmp(p+1,facys[i].name)==0)
					conf->facility=facys[i].fac;
		#endif
                }
        }

fclose(fic);

return(0);
}

void Log(config *conf,char *recipient,char *sender,char *ip,int white)
{
#ifdef HAVE_SYSLOG_H
openlog("gld",0,conf->facility);
if(white==MSGGREYLIST) syslog(LOG_NOTICE,"Greylist activated for recipient=<%s> sender=<%s> ip=<%s>",recipient,sender,ip);
if(white==MSGLOCALWL) syslog(LOG_NOTICE,"Local whitelist hit for recipient=<%s> sender=<%s> ip=<%s>",recipient,sender,ip);
if(white==MSGDNSWL) syslog(LOG_NOTICE,"DNS whitelist hit for recipient=<%s> sender=<%s> ip=<%s>",recipient,sender,ip);
closelog();
#endif
}

void ErrorLog(config *conf,char *msg)
{
#ifdef HAVE_SYSLOG_H
openlog("gld",0,conf->facility);
syslog(LOG_ALERT,"%s",msg);
closelog();
#endif
}

int ReadClients(config *conf,char *str)
{
char *ptr,*x,*y;

ptr=str;
conf->nbnet=0;

while(*ptr!=0)
        {
        x=strstr(ptr," ");
        if(x!=NULL) *x=0;
        y=strstr(ptr,"/");
        if(y!=NULL)
                {
                *y=0;
                conf->nets[conf->nbnet].netw=Hash(ptr);
                conf->nets[conf->nbnet].mask=CidrMsk(atoi(y+1));
                conf->nbnet++;
                if(conf->nbnet==NLEN || x==NULL) break;
                }
        ptr=x+1;
        }
return(0);
}

unsigned long Hash(char *ip)
{
struct in_addr x;
int r;

r=inet_aton(ip,&x);
if(r==0) return(0);

return(x.s_addr);
}

int CheckIP(config *conf,char *ip)
{
unsigned long h,netw,mask;
int i;

if(conf->nbnet==0) return(1);

h=Hash(ip);

for(i=0;i<conf->nbnet;i++)
        {
        netw=conf->nets[i].netw;
        mask=conf->nets[i].mask;
        if((h&mask)==(netw&mask)) return(1);
        }

return(0);
}

unsigned long CidrMsk(int msk)
{
int i;
unsigned long x=0;

for(i=0;i<msk;i++)
        {
        x=x>>1;
        x+=0x80000000;
        }

x=htonl(x);
return(x);
}
